### PR TITLE
Support pthread mode for ExecutionQueue

### PR DESCRIPTION
**Pull request by chenBright**
### What problem does this PR solve?

Issue Number:

Problem Summary:

### What is changed and the side effects?

Changed:

ExecutionQueue支持使用pthread来执行任务。

Side effects:
- Performance effects(性能影响):

- Breaking backward compatibility(向后兼容性): 

---
### Check List:
- Please make sure your changes are compilable(请确保你的更改可以通过编译).
- When providing us with a new feature, it is best to add related tests(如果你向我们增加一个新的功能, 请添加相关测试).
- Please follow [Contributor Covenant Code of Conduct](../../master/CODE_OF_CONDUCT.md).(请遵循贡献者准则).

### CODE REVIEW

File: src/bthread/execution_queue.cpp [hunk 1]
```diff
@@ -189,6 +208,28 @@ void* ExecutionQueueBase::_execute_tasks(void* arg) {
...
 
+void* ExecutionQueueBase::_execute_tasks_pthread(void* arg) {
+    butil::PlatformThread::SetName("ExecutionQueue");
+    auto head = (TaskNode*)arg;
+    auto m = (ExecutionQueueBase*)head->q;
+    m->_current_head = head;
+    while (true) {
+        BAIDU_SCOPED_LOCK(m->_mutex);
+        while (!m->_current_head) {
+            m->_cond.Wait();
```
Comment (wasphin) [inline src/bthread/execution_queue.cpp:219]: 这里 join 时有可能 block 住吗？

Comment (chenBright) [reply src/bthread/execution_queue.cpp:219]: 不会。stop之后，_execute_tasks_pthread不会block在条件变量上，而是在执行完stop task后就会跳出循环并退出线程了。

File: src/bthread/execution_queue_inl.h [hunk 2]
```diff
@@ -234,6 +238,13 @@ friend class TaskIteratorBase;
     clear_task_mem _clear_func;
     ExecutionQueueOptions _options;
     butil::atomic<int>* _join_butex;
+
+    // For pthread mode.
+    pthread_t _pid;
+    bool _pthread_started;
+    butil::Mutex _mutex;
```
Comment (wwbmmm) [inline src/bthread/execution_queue_inl.h:245]: 能不用锁吗。原来ExecutionQueue用的是无锁算法，加上锁可能会影响性能

Comment (chenBright) [reply src/bthread/execution_queue_inl.h:245]: 从ExecutionQueueTest.performance结果来看，对性能是没影响的吧。
```shell
[ RUN      ] ExecutionQueueTest.performance
I0729 10:30:21.869438  9911 bthread_execution_queue_unittest.cpp:217] ================ bthread ================
PROFILE: interrupts/evictions/bytes = 92/8/2992
I0729 10:30:23.000201  9911 bthread_execution_queue_unittest.cpp:240] With addressed execq, each execution_queue_execute takes 957 total_num=4078814 ns with 8 threads
PROFILE: interrupts/evictions/bytes = 100/1/6416
I0729 10:30:23.881414  9911 bthread_execution_queue_unittest.cpp:266] With id explicitly, execution_queue_execute takes 1578 total_num=2505991 ns with 8 threads
I0729 10:30:23.881448  9911 bthread_execution_queue_unittest.cpp:215] ================ pthread ================
PROFILE: interrupts/evictions/bytes = 55/7/2568
I0729 10:30:24.918703  9911 bthread_execution_queue_unittest.cpp:240] With addressed execq, each execution_queue_execute takes 591 total_num=3632640 ns with 8 threads
PROFILE: interrupts/evictions/bytes = 96/1/5576
I0729 10:30:25.824709  9911 bthread_execution_queue_unittest.cpp:266] With id explicitly, execution_queue_execute takes 1562 total_num=2530991 ns with 8 threads
[       OK ] ExecutionQueueTest.performance (3955 ms)
```

**Comment by wwbmmm** [comment]
LGTM

File: src/bthread/execution_queue.cpp [hunk 3]
```diff
@@ -105,16 +106,34 @@ void ExecutionQueueBase::start_execute(TaskNode* node) {
...
-            PLOG(FATAL) << "Fail to start bthread";
-            _execute_tasks(node);
+        if (_options.use_pthread) {
+            if (_pthread_started) {
+                BAIDU_SCOPED_LOCK(_mutex);
+                _current_head = node;
+                _cond.Signal();
+            } else {
+                // Start the execution bthread in background once.
+                if (pthread_create(&_pid, NULL,
```
Comment (legionxiong) [inline src/bthread/execution_queue.cpp:116]: Why not create a thread pool?

Comment (legionxiong) [reply src/bthread/execution_queue.cpp:116]: This may cause hundreds of threads being created.

Comment (chenBright) [reply src/bthread/execution_queue.cpp:116]: A thread pool shared by multiple ExecutionQueues may affect performance.

Users should use bthread mode first. Pthread mode should only be used in specific scenarios.

**Comment by chenBright** [comment]
没有其他问题的话，这个PR应该可以合了

**Pull request merged**